Description: Change default terminal command to 'x-terminal-emulator'
Author: Graham Inggs <graham@nerve.org.za>
Forwarded: not-needed
Last-Update: 2015-02-15
--- a/src/platform/uOSUtils.pas
+++ b/src/platform/uOSUtils.pas
@@ -53,8 +53,8 @@
   fmtRunInTerm = '';
   MonoSpaceFont = 'Monaco';
   {$ELSE}
-  RunTerm = 'xterm';  // default terminal
-  RunInTerm = 'xterm -e sh -c'; // default run in terminal command
+  RunTerm = 'x-terminal-emulator';  // default terminal
+  RunInTerm = 'x-terminal-emulator -e sh -c'; // default run in terminal command
   fmtRunInTerm = '%s ''%s ; echo -n Press ENTER to exit... ; read a''';
   MonoSpaceFont = 'Monospace';
   {$ENDIF}
