Index: AbZBrows.pas
===================================================================
--- AbZBrows.pas	(revision 512)
+++ AbZBrows.pas	(working copy)
@@ -129,7 +129,7 @@
 implementation
 
 uses
-  SysUtils, AbBzip2Typ, AbExcept, AbGzTyp, AbTarTyp, AbUtils;
+  SysUtils, AbBzip2Typ, AbExcept, AbGzTyp, AbTarTyp, AbUtils, DCOSUtils;
 
 { TAbCustomZipBrowser implementation ======================================= }
 
@@ -205,7 +205,7 @@
   FArchive := nil;
 
   if FileName <> '' then begin
-    if FileExists(FileName) then begin { open it }
+    if mbFileExists(FileName) then begin { open it }
       ArcType := ArchiveType;
       if not ForceType then
          ArcType := AbDetermineArcType(FileName, atUnknown);
Index: AbZipKit.pas
===================================================================
--- AbZipKit.pas	(revision 512)
+++ AbZipKit.pas	(working copy)
@@ -93,6 +93,8 @@
       {extract all tagged items from the archive}
     procedure ExtractToStream(const aFileName : string; ToStream : TStream);
       {extract the specified item to TStream descendant}
+    procedure TestItemAt(Index : Integer);
+      {test specific item in the archive}
     procedure TestTaggedItems;
       {test all tagged items in the archive}
 
@@ -253,6 +255,14 @@
     (ZipArchive as TAbZipArchive).PasswordRetries := Value;
 end;
 { -------------------------------------------------------------------------- }
+procedure TAbCustomZipKit.TestItemAt(Index : Integer);
+begin
+  if (ZipArchive <> nil) then
+    ZipArchive.TestItemAt(Index)
+  else
+    raise EAbNoArchive.Create;
+end;
+{ -------------------------------------------------------------------------- }
 procedure TAbCustomZipKit.TestTaggedItems;
   {test all tagged items in the archive}
 begin
