Index: AbZipKit.pas
===================================================================
--- AbZipKit.pas	(revision 512)
+++ AbZipKit.pas	(working copy)
@@ -93,6 +93,8 @@
       {extract all tagged items from the archive}
     procedure ExtractToStream(const aFileName : string; ToStream : TStream);
       {extract the specified item to TStream descendant}
+    procedure TestItemAt(Index : Integer);
+      {test specific item in the archive}
     procedure TestTaggedItems;
       {test all tagged items in the archive}
 
@@ -253,6 +255,14 @@
     (ZipArchive as TAbZipArchive).PasswordRetries := Value;
 end;
 { -------------------------------------------------------------------------- }
+procedure TAbCustomZipKit.TestItemAt(Index : Integer);
+begin
+  if (ZipArchive <> nil) then
+    ZipArchive.TestItemAt(Index)
+  else
+    raise EAbNoArchive.Create;
+end;
+{ -------------------------------------------------------------------------- }
 procedure TAbCustomZipKit.TestTaggedItems;
   {test all tagged items in the archive}
 begin
