Index: interfaces/qt/qt45.pas
===================================================================
--- interfaces/qt/qt45.pas	(revision 47328)
+++ interfaces/qt/qt45.pas	(working copy)
@@ -12411,7 +12411,7 @@
 procedure QGraphicsView_invalidateScene(handle: QGraphicsViewH; rect: QRectFH = nil; layers: QGraphicsSceneSceneLayers = QGraphicsSceneAllLayers); cdecl; external Qt4PasLib name 'QGraphicsView_invalidateScene';
 procedure QGraphicsView_updateSceneRect(handle: QGraphicsViewH; rect: QRectFH); cdecl; external Qt4PasLib name 'QGraphicsView_updateSceneRect';
 
-
+{
 type
   QSslKeyType = ( // QSsl::KeyType (1)
     QSslPrivateKey, QSslPublicKey );
@@ -13284,7 +13284,7 @@
 function QLCLWebView_create(parent: QWidgetH = nil): QLCLWebViewH; cdecl; external Qt4PasLib name 'QLCLWebView_create';
 procedure QLCLWebView_destroy(handle: QLCLWebViewH); cdecl; external Qt4PasLib name 'QLCLWebView_destroy'; 
 procedure QLCLWebView_override_createWindow(handle: QLCLWebViewH; hook: QLCLWebView_createWindow_Override); cdecl; external Qt4PasLib name 'QLCLWebView_override_createWindow';
-
+}
 function QCoreApplication_hook_create(handle: QObjectH): QCoreApplication_hookH; cdecl; external Qt4PasLib name 'QCoreApplication_hook_create';
 procedure QCoreApplication_hook_destroy(handle: QCoreApplication_hookH); cdecl; external Qt4PasLib name 'QCoreApplication_hook_destroy'; 
 procedure QCoreApplication_hook_hook_aboutToQuit(handle: QCoreApplication_hookH; hook: QCoreApplication_aboutToQuit_Event); cdecl; external Qt4PasLib name 'QCoreApplication_hook_hook_aboutToQuit';
@@ -13797,7 +13797,7 @@
 procedure QGraphicsScene_hook_destroy(handle: QGraphicsScene_hookH); cdecl; external Qt4PasLib name 'QGraphicsScene_hook_destroy'; 
 procedure QGraphicsScene_hook_hook_sceneRectChanged(handle: QGraphicsScene_hookH; hook: QGraphicsScene_sceneRectChanged_Event); cdecl; external Qt4PasLib name 'QGraphicsScene_hook_hook_sceneRectChanged';
 procedure QGraphicsScene_hook_hook_selectionChanged(handle: QGraphicsScene_hookH; hook: QGraphicsScene_selectionChanged_Event); cdecl; external Qt4PasLib name 'QGraphicsScene_hook_hook_selectionChanged';
-
+{
 function QAbstractSocket_hook_create(handle: QObjectH): QAbstractSocket_hookH; cdecl; external Qt4PasLib name 'QAbstractSocket_hook_create';
 procedure QAbstractSocket_hook_destroy(handle: QAbstractSocket_hookH); cdecl; external Qt4PasLib name 'QAbstractSocket_hook_destroy'; 
 procedure QAbstractSocket_hook_hook_hostFound(handle: QAbstractSocket_hookH; hook: QAbstractSocket_hostFound_Event); cdecl; external Qt4PasLib name 'QAbstractSocket_hook_hook_hostFound';
@@ -13898,9 +13898,9 @@
 procedure QWebView_hook_hook_selectionChanged(handle: QWebView_hookH; hook: QWebView_selectionChanged_Event); cdecl; external Qt4PasLib name 'QWebView_hook_hook_selectionChanged';
 procedure QWebView_hook_hook_iconChanged(handle: QWebView_hookH; hook: QWebView_iconChanged_Event); cdecl; external Qt4PasLib name 'QWebView_hook_hook_iconChanged';
 procedure QWebView_hook_hook_urlChanged(handle: QWebView_hookH; hook: QWebView_urlChanged_Event); cdecl; external Qt4PasLib name 'QWebView_hook_hook_urlChanged';
+}
 
 
-
 //=======================================================
 // Dynamic Qt Version 
 //=======================================================
