Index: Qt4Pas.pro
===================================================================
--- Qt4Pas.pro	(revision 1)
+++ Qt4Pas.pro	(working copy)
@@ -12,7 +12,6 @@
 
 VERSION = 5.2.5
 
-QT += network webkit
 TARGET = Qt4Pas
 TEMPLATE = lib
 DEPENDPATH += src
@@ -155,21 +154,7 @@
            qprintdialog_hook.h \
            qprintpreviewdialog_hook.h \
            qsystemtrayicon_hook.h \
-           qgraphicsscene_hook.h \
-           qabstractsocket_hook.h \
-           qudpsocket_hook.h \
-           qtcpsocket_hook.h \
-           qtcpserver_hook.h \
-           qnetworkaccessmanager_hook.h \
-           qnetworkreply_hook.h \
-           qwebframe_hook.h \
-           qwebsecurityorigin_hook.h \
-           qwebdatabase_hook.h \
-           qwebhistory_hook.h \
-           qwebhistoryinterface_hook.h \
-           qwebpage_hook.h \
-           qwebsettings_hook.h \
-           qwebview_hook.h
+           qgraphicsscene_hook.h
 
 SOURCES += qt4pas.cpp \
            qobject_hook_c.cpp \
@@ -335,32 +320,6 @@
            qstylefactory_c.cpp \
            qgraphicsscene_c.cpp \
            qgraphicsview_c.cpp \
-           qsslcipher_c.cpp \
-           qsslkey_c.cpp \
-           qsslerror_c.cpp \
-           qabstractsocket_c.cpp \
-           qudpsocket_c.cpp \
-           qtcpsocket_c.cpp \
-           qtcpserver_c.cpp \
-           qsslconfiguration_c.cpp \
-           qsslsocket_c.cpp \
-           qnetworkaccessmanager_c.cpp \
-           qnetworkrequest_c.cpp \
-           qnetworkreply_c.cpp \
-           qnetworkcookie_c.cpp \
-           qlclnetworkcookie_c.cpp \
-           qnetworkproxy_c.cpp \
-           qauthenticator_c.cpp \
-           qwebframe_c.cpp \
-           qwebsecurityorigin_c.cpp \
-           qwebdatabase_c.cpp \
-           qwebhistory_c.cpp \
-           qwebhistoryinterface_c.cpp \
-           qwebpage_c.cpp \
-           qlclwebpage_c.cpp \
-           qwebsettings_c.cpp \
-           qwebview_c.cpp \
-           qlclwebview_c.cpp \
            qcoreapplication_hook_c.cpp \
            qtimer_hook_c.cpp \
            qabstractitemmodel_hook_c.cpp \
@@ -434,20 +393,6 @@
            qprintdialog_hook_c.cpp \
            qprintpreviewdialog_hook_c.cpp \
            qsystemtrayicon_hook_c.cpp \
-           qgraphicsscene_hook_c.cpp \
-           qabstractsocket_hook_c.cpp \
-           qudpsocket_hook_c.cpp \
-           qtcpsocket_hook_c.cpp \
-           qtcpserver_hook_c.cpp \
-           qnetworkaccessmanager_hook_c.cpp \
-           qnetworkreply_hook_c.cpp \
-           qwebframe_hook_c.cpp \
-           qwebsecurityorigin_hook_c.cpp \
-           qwebdatabase_hook_c.cpp \
-           qwebhistory_hook_c.cpp \
-           qwebhistoryinterface_hook_c.cpp \
-           qwebpage_hook_c.cpp \
-           qwebsettings_hook_c.cpp \
-           qwebview_hook_c.cpp
+           qgraphicsscene_hook_c.cpp
 
 # end of file
\ No newline at end of file
Index: src/qt4pas.cpp
===================================================================
--- src/qt4pas.cpp	(revision 1)
+++ src/qt4pas.cpp	(working copy)
@@ -15,8 +15,6 @@
 
 #include <QtCore>
 #include <QtGui>
-#include <QtNetwork>
-#include <QtWebKit>
 
 
 #include "qobject_hook_c.cpp"
@@ -182,32 +180,6 @@
 #include "qstylefactory_c.cpp"
 #include "qgraphicsscene_c.cpp"
 #include "qgraphicsview_c.cpp"
-#include "qsslcipher_c.cpp"
-#include "qsslkey_c.cpp"
-#include "qsslerror_c.cpp"
-#include "qabstractsocket_c.cpp"
-#include "qudpsocket_c.cpp"
-#include "qtcpsocket_c.cpp"
-#include "qtcpserver_c.cpp"
-#include "qsslconfiguration_c.cpp"
-#include "qsslsocket_c.cpp"
-#include "qnetworkaccessmanager_c.cpp"
-#include "qnetworkrequest_c.cpp"
-#include "qnetworkreply_c.cpp"
-#include "qnetworkcookie_c.cpp"
-#include "qlclnetworkcookie_c.cpp"
-#include "qnetworkproxy_c.cpp"
-#include "qauthenticator_c.cpp"
-#include "qwebframe_c.cpp"
-#include "qwebsecurityorigin_c.cpp"
-#include "qwebdatabase_c.cpp"
-#include "qwebhistory_c.cpp"
-#include "qwebhistoryinterface_c.cpp"
-#include "qwebpage_c.cpp"
-#include "qlclwebpage_c.cpp"
-#include "qwebsettings_c.cpp"
-#include "qwebview_c.cpp"
-#include "qlclwebview_c.cpp"
 #include "qcoreapplication_hook_c.cpp"
 #include "qtimer_hook_c.cpp"
 #include "qabstractitemmodel_hook_c.cpp"
@@ -282,17 +254,3 @@
 #include "qprintpreviewdialog_hook_c.cpp"
 #include "qsystemtrayicon_hook_c.cpp"
 #include "qgraphicsscene_hook_c.cpp"
-#include "qabstractsocket_hook_c.cpp"
-#include "qudpsocket_hook_c.cpp"
-#include "qtcpsocket_hook_c.cpp"
-#include "qtcpserver_hook_c.cpp"
-#include "qnetworkaccessmanager_hook_c.cpp"
-#include "qnetworkreply_hook_c.cpp"
-#include "qwebframe_hook_c.cpp"
-#include "qwebsecurityorigin_hook_c.cpp"
-#include "qwebdatabase_hook_c.cpp"
-#include "qwebhistory_hook_c.cpp"
-#include "qwebhistoryinterface_hook_c.cpp"
-#include "qwebpage_hook_c.cpp"
-#include "qwebsettings_hook_c.cpp"
-#include "qwebview_hook_c.cpp"
