Description: Do not display build info in window title
Author: Graham Inggs <graham@nerve.org.za>
Forwarded: not-needed
Last-Update: 2015-02-15
--- a/src/fmain.pas
+++ b/src/fmain.pas
@@ -964,13 +964,7 @@
     if Length(UniqueInstance.ServernameByUser) > 0 then
       ServernameString := ' [' + UniqueInstance.ServernameByUser + ']';
 
-    Result := Format('%s%s %s build %s; %s',
-        ['Double Commander',
-        ServernameString,
-        dcVersion,
-        dcRevision,
-        dcBuildDate]
-    );
+    Result := 'Double Commander';
   end;
 
 var
