Description: Change default terminal command to 'x-terminal-emulator'
Author: Graham Inggs <ginggs@debian.org>
Forwarded: not-needed
Last-Update: 2016-03-14
--- a/src/platform/uOSUtils.pas
+++ b/src/platform/uOSUtils.pas
@@ -65,11 +65,11 @@
   RunInTermCloseParams = '';
   MonoSpaceFont = 'Monaco';
   {$ELSE}
-  RunTermCmd = 'xterm';  // default terminal
+  RunTermCmd = 'x-terminal-emulator';  // default terminal
   RunTermParams = '';
-  RunInTermStayOpenCmd = 'xterm'; // default run in terminal command AND Stay open after command
+  RunInTermStayOpenCmd = 'x-terminal-emulator'; // default run in terminal command AND Stay open after command
   RunInTermStayOpenParams = '-e sh -c ''{command}; echo -n Press ENTER to exit... ; read a''';
-  RunInTermCloseCmd = 'xterm'; // default run in terminal command AND Close after command
+  RunInTermCloseCmd = 'x-terminal-emulator'; // default run in terminal command AND Close after command
   RunInTermCloseParams = '-e sh -c {command}';
   MonoSpaceFont = 'Monospace';
   {$ENDIF}
